### ISSUE TITLE
3D implementation

### ISSUE THREAD

**Issue by abarret** [abarret/multiphase-stokes#39]
At some point, we will need to implement a 3D solver for these equations. There are a few things that will need to be updated. The main issue is the preconditioner probably won't scale to 3D. We need to switch away from cell-wise Vanka smoothers to larger box Vanka smoothers.

- [ ] 3D implementation of the operator
- [ ] Construct PETSc matrices for the smoothing operators (use e.g. ILU)
- [ ] Use a PETSc solver for the coarsest level solve.

**Comment by abarret** [comment]
@bindi-nagda it would be very helpful for this if you wrote down the details of the smoother discretization somewhere (maybe your dissertation?)

**Comment by bindi-nagda** [comment]
Yeah, I've included the details of the smoother discretization in my dissertation. I can send you a copy of it.

**Comment by abarret** [comment]
Following up on this, initial tests indicate that we spend almost 90% of our time solving the linear system in the smoother. This suggests that this preconditioner definitely won't scale to 3D, as it will only get more expensive (solving a 13 by 13 system on each grid cell as opposed to a 9 by 9).

**Comment by abarret** [comment]
Because we spend almost 90% of our time in the smoother, we may be able to significantly decrease run time by implementing our own LU or SVD solver in the smoother routine instead of using the lapack implementation. The compiler would then be free to optimize the matrix solve because the system size will be known at compile time.

**Comment by bindi-nagda** [comment]
> Because we spend almost 90% of our time in the smoother, we may be able to significantly decrease run time by implementing our own LU or SVD solver in the smoother routine instead of using the lapack implementation. The compiler would then be free to optimize the matrix solve because the system size will be known at compile time.

Sound good. Can we talk about this on Friday at our meeting so I can try to implement it?

**Comment by abarret** [comment]
> Sound good. Can we talk about this on Friday at our meeting so I can try to implement it?

Sure. We can discuss it on Friday. My suggestion is to finish up convergence testing so we can push out a paper before you try this.
